In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('hotel_booking.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

# Stage 2 Data Preprocessing

[Stage 2] Final Project DS 23

Periode : 15 - 21 Agustus

Halo teman-teman! Setelah menemukan insight, dan visualiasi dari dataset yang dipilih, kini saatnya teman-teman masuk ke dalam proses Data Pre-Processing untuk menentukan :
1. Apakah sudah melakukan pengecekkan data bermasalah seperti missing values, invalid values, atau data duplicate dan sudah membersihkannya?
2. Apakah sudah menentukan feature apa saja yang akan digunakan, atau perlu ditambahkan, dan reformatting feature sesuai dengan kebutuhan?

Penting! Saat memanipulasi data, mohon dilakukan dengan alasan yang jelas, dan tidak melakukan penambahan feature baru tapi tidak ada alasan yang mendasari langkah tersebut diperlukan.

Silahkan selesaikan dan kumpulkan dokumen Stage 2 tersebut (Notulen Mentoring, Source Code, Laporan Project) ke link gdrive yang sudah disiapkan (akan diinfokan secara personal ke masing-masing tim), paling lambat tanggal 21 Agustus 2022 pukul 23:59

# Buat Dataset untuk hasil Preprocessing

In [6]:
df_clean = df.copy()
df_clean.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

# Hotel

Perlu dilakukan OHE pada kolom Hotel dengan merubah setiap nilai kolom menjadi kolom baru dengan nilai biner yaitu 0 atau 1. Penggunaan OHE ini karena kolom Hotel tidak memiliki urutan/tingkatan. Dimana Resort hotel dilambangkan 1 dan City hotel dilambangkan 0.

In [7]:
# OHE
hotel_ohe = pd.get_dummies(df.hotel,drop_first=True)
# Gabungkan hasil OHE ke dataset
df_clean = df_clean.join(hotel_ohe)
# Hapus kolom `hotel`
df_clean = df_clean.drop('hotel', axis=1)


df_clean.head(1)

is_canceled  lead_time  arrival_date_year arrival_date_month  \
0            0        342               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  adr  \
0                        0                     0       2       0.0  ...  0.0   

  required_car_parking_spaces total_of_special_requests reservation_status  \
0                           0                         0          Check-Out   

  reservation_status_date           name                        email  \
0              2015-07-01  Ernest Barnes  Ernest.Barnes31@outlook.com   

   phone-number       credit_card Resort Hotel  
0  669-792-1661  ************4322            1  

[1 rows x 36 columns]

# Lead Time

Dibagi 30, untuk mendapatkan waktu tunggu dalam bulan

In [8]:
df_clean['month_lead_time'] = df_clean['lead_time'].div(30).round()
df_clean.head(1)

#hapus kolom `lead_time`
# df_clean = df_clean.drop('lead_time', axis=1)

is_canceled  lead_time  arrival_date_year arrival_date_month  \
0            0        342               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  \
0                        0                     0       2       0.0  ...   

   required_car_parking_spaces total_of_special_requests reservation_status  \
0                            0                         0          Check-Out   

  reservation_status_date           name                        email  \
0              2015-07-01  Ernest Barnes  Ernest.Barnes31@outlook.com   

   phone-number       credit_card Resort Hotel month_lead_time  
0  669-792-1661  ************4322            1            11.0  

[1 rows x 37 columns]

# Arrival Date

## arrival_date_year, arrival_date_month, dan arrival_date_day_of_month

Membuat kolom `date` yang merupakan gabungan dari kolom `arrival_date_year`, `arrival_date_month`, dan `arrival_date_day_of_month`

Membuat kolom `is_weekend` untuk mendapatkan informasi mengenai tanggal arrival apakah di weekend atau weekdays

In [9]:
def month(x):
    if x['arrival_date_month'] == 'January':
            month = '1'
    elif x['arrival_date_month'] == 'February':
            month = '2'
    elif x['arrival_date_month'] == 'March':
            month = '3' 
    elif x['arrival_date_month'] =='April':
            month = '4'
    elif x['arrival_date_month'] == 'May':
            month = '5'    
    elif x['arrival_date_month'] == 'June':
            month = '6'
    elif x['arrival_date_month'] == 'July':
            month = '7'
    elif x['arrival_date_month'] == 'August':
            month = '8'
    elif x['arrival_date_month'] == 'September':
            month = '9'
    elif x['arrival_date_month'] == 'October':
            month = '10'
    elif x['arrival_date_month'] == 'November':
            month = '11'
    else:
            month = '12'
    return month

In [10]:
df_clean['month_number'] = df_clean.apply(lambda x: month(x), axis=1)
# merubah tahun dan hari menjadi string
df_clean["arrival_date_day_of_month"] = df_clean["arrival_date_day_of_month"].astype(str)
df_clean["arrival_date_year"] = df_clean["arrival_date_year"].astype(str)
# membuat kolom date of arrival
df_clean["date"] = pd.to_datetime(df_clean["arrival_date_year"] + "/" + df_clean["month_number"] + "/" + df_clean["arrival_date_day_of_month"])  

# menambahkan info tanggal weekend
df_clean["isWeekend"] = df_clean["date"].dt.weekday >= 5
def weekend(x):
    if x['isWeekend'] ==True:
            weekend = '1'
    else: 
            weekend = '0'
    return weekend
df_clean['is_weekend'] = df_clean.apply(lambda x: weekend(x), axis=1)


In [11]:
# Menghapus kolom yang tidak terpakai
df_clean = df_clean.drop(['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'isWeekend'], axis=1)

df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights  adults  children  babies meal country  ...  \
0                     0       2       0.0       0   BB     PRT  ...   

  reservation_status_date           name                        email  \
0              2015-07-01  Ernest Barnes  Ernest.Barnes31@outlook.com   

   phone-number       credit_card Resort Hotel month_lead_time  month_number  \
0  669-792-1661  ************4322            1            11.0             7   

        date  is_weekend  
0 2015-07-01           0  

[1 rows x 37 columns]

## arrival_date_week_number

Kolom `arrival_date_week_number` tidak diubah

In [12]:
# df_clean = df_clean.drop('arrival_date_week_number', axis=1)

# Stays

Akan dibuat kolom total_stay dari gabungan kolom stays_in_weekend_nights dan stays_in_week_nights

In [ ]:
df_clean['total_stay'] = df_clean.apply(lambda x: (x['stays_in_weekend_nights']) + (x['stays_in_week_nights']), axis=1)
#hapus kolom stays_weekend dan weeknight
df_clean = df_clean.drop(['stays_in_weekend_nights','stays_in_week_nights'],axis=1)

# Adults, children, dan babies

## Adults

In [13]:
df.adults.describe()

count    119390.000000
mean          1.856403
std           0.579261
min           0.000000
25%           2.000000
50%           2.000000
75%           2.000000
max          55.000000
Name: adults, dtype: float64

Kolom adults memiliki nilai minimal 0, ini sebuah kesalahan dimana booking harus ada orang dewasa

In [14]:
df[df['adults']<1].shape

(403, 36)

Ada 403 baris yang memiliki nilai < 1

Drop baris dengan nilai dibawah 1

In [15]:
df_clean = df_clean[df_clean['adults']>0]
df_clean.shape

(118987, 37)

## Children, babies

Dilakukan penjumlahan antara kolom children dan babies menjadi kolom Kids, karena diasumsikan children dan babies masuk ke dalam kategori yang sama, yaitu kids. 

In [16]:
#Kolom children yang nilainya kosong diganti dengan 0.
df_clean.children = df_clean.children.fillna(0)

#Kolom kids (children dan babies)
df_clean['kids'] = df_clean.apply(lambda x: (x['children']) + (x['babies']), axis=1)

# Mengubah tipe data menjadi int
df_clean.kids = df_clean.kids.astype(int)

## Buat kolom baru Total guest dan is_have_kids

Buat kolom baru :
- `total_guest` : gabungan dari kolom `adults` dan `kids`
- `is_have_kids` : status apakah memiliki anak-anak atau tidak

In [17]:
df_clean['total_guest'] = df_clean['adults'] + df_clean['kids']

df_clean['is_have_kids']  = df_clean['kids'].apply(lambda x : 0 if x<1 else 1)

In [18]:
# Menghapus kolom yang tidak diperlukan
df_clean = df_clean.drop(['adults', 'children', 'babies', 'kids'], axis=1)

In [19]:
df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights meal country market_segment distribution_channel  \
0                     0   BB     PRT         Direct               Direct   

   is_repeated_guest  ...                        email  phone-number  \
0                  0  ...  Ernest.Barnes31@outlook.com  669-792-1661   

        credit_card Resort Hotel  month_lead_time month_number       date  \
0  ************4322            1             11.0            7 2015-07-01   

   is_weekend  total_guest is_have_kids  
0           0            2            0  

[1 rows x 36 columns]

# Meal

Perlu dilakukan OHE pada kolom meal dengan merubah setiap nilai kolom menjadi kolom baru dengan nilai biner yaitu 0 atau 1. Penggunaan OHE ini karena kolom meal tidak memiliki urutan/tingkatan.

Value :
- BB : bed and breakfast
- HB : half board (breakfast and dinner included)
- FB : full board (breakfast, lunch, and dinner included)
- SC : self catering (no meals included)
- Undefined

In [20]:
df_clean.meal.value_counts()

BB           92020
HB           14454
SC           10546
Undefined     1169
FB             798
Name: meal, dtype: int64

Ubah value Undefined ke SC (no meals)

In [21]:
df_clean['meal'] = df_clean['meal'].apply(lambda x: 'SC' if x=='Undefined' else x)

In [22]:
df_clean.meal.value_counts()

BB    92020
HB    14454
SC    11715
FB      798
Name: meal, dtype: int64

In [23]:
meal_ohe = pd.get_dummies(df_clean.meal,drop_first=True, prefix='meal')
df_clean = df_clean.join(meal_ohe)

In [24]:
df_clean = df_clean.drop('meal', axis=1)

# Country

Dikarenakan terdapat banyak data country yang memungkinkan apabila digunakan untuk analisis, jumlah data negara PRT (Portugal) ... maka dikelompokkan data country menjadi 2 yaitu Portugal dan Non Portugal. Dimana Portugal diberi label 1 dan Non Portugal diberi label 0.

In [25]:
df_clean['country'].isnull().sum()

478

Mengubah nilai kosong pada kolom Country menjadi PRT

In [26]:
df_clean['country'] = df_clean['country'].fillna('PRT')

In [27]:
df_clean['country'].isnull().sum()

0

In [28]:
#Kolom country
df_clean['negara'] = df_clean.apply(lambda x: "Portugal" if (x['country'] == 'PRT') else "Others", axis=1)

In [29]:
df_clean.negara.value_counts()

Others      70069
Portugal    48918
Name: negara, dtype: int64

In [30]:
mapping_negara = {
    'Portugal' : 1,
    'Others' : 0,
}

Buat kolom `is_lokal` yang merupakan status apabila berasal dari lokal (portugal) atau tidak

In [31]:
df_clean['is_lokal'] = df_clean['negara'].map(mapping_negara)


In [32]:
# Hapus kolom Country
df_clean = df_clean.drop(['country', 'negara'], axis=1)

df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
0                     0     PRT         Direct               Direct   

   is_repeated_guest  previous_cancellations  ...  month_number       date  \
0                  0                       0  ...             7 2015-07-01   

  is_weekend  total_guest is_have_kids  meal_FB  meal_HB  meal_SC    negara  \
0          0            2            0        0        0        0  Portugal   

   is_lokal  
0         1  

[1 rows x 40 columns]

# Market_Segment

Kolom Market_Segment akan di drop karena redundant dengan kolom Distribution_Channel

In [ ]:
df_clean = df_clean.drop(['market_Segment'],axis=1)

# Distribution Channel

Di OHE

# Is_repeated_guest

Kolom `is_repeated_guest` tidak diubah

In [33]:
df.is_repeated_guest.value_counts()

0    115580
1      3810
Name: is_repeated_guest, dtype: int64

# previous_cancellations

In [34]:
df.previous_cancellations.value_counts()

0     112906
1       6051
2        116
3         65
24        48
11        35
4         31
26        26
25        25
6         22
19        19
5         19
14        14
13        12
21         1
Name: previous_cancellations, dtype: int64

Akan diubah valuenya:
- 0 > tidak pernah cancel
- 1 > pernah cancel 1x atau lebih

In [35]:
df_clean['previous_cancellations'] = df_clean['previous_cancellations'].apply(lambda x: 0 if x == 0 else 1)
df_clean['previous_cancellations'].value_counts()

0    112510
1      6477
Name: previous_cancellations, dtype: int64

# previous_bookings_not_canceled

Akan dihapus karena redundant dengan `previous_cancellations`

In [36]:
df.previous_bookings_not_canceled.value_counts()

0     115770
1       1542
2        580
3        333
4        229
       ...  
47         1
49         1
50         1
51         1
72         1
Name: previous_bookings_not_canceled, Length: 73, dtype: int64

In [37]:
df_clean = df_clean.drop('previous_bookings_not_canceled', axis=1)

# Reserved_room_type dan assigned_room_type

Buat kolom `room_as_ordered` dari perbandingan kolom `reserved_room_type` dan `assigned_room_type`

In [38]:
df_clean['room_as_ordered'] = np.where(df_clean['reserved_room_type'] == df_clean['assigned_room_type'], 1, 0)

# Hapus kolom tidak terpakai
df_clean.drop(['reserved_room_type', 'assigned_room_type'], axis=1)

df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
0                     0     PRT         Direct               Direct   

   is_repeated_guest  previous_cancellations  ...       date is_weekend  \
0                  0                       0  ... 2015-07-01          0   

   total_guest is_have_kids  meal_FB  meal_HB  meal_SC    negara  is_lokal  \
0            2            0        0        0        0  Portugal         1   

   room_as_ordered  
0                1  

[1 rows x 40 columns]

# booking_changes

- Kolom booking_changes memiliki value numerical, akan tetapi untuk memudahkan maka diubah medengan menggunakan conditions dengan asumsi:
- Jika booking tidak ada perubahan makan = 0 maka dikategorikan sebagai 0, sedangkan jika customer perubahan booking >= 1 maka dikategorikan sebagai 1.

In [ ]:
df_clean['booking_changes'] = [0 if x == 0 else 1 for x in df_clean['booking_changes']]

# Deposit Type

Perlu dilakukan OHE pada kolom deposit_type dengan merubah setiap nilai kolom menjadi kolom baru dengan nilai biner yaitu 0 atau 1. Penggunaan OHE ini karena deposit_type dan tidak memiliki urutan/tingkatan.

In [39]:
deposit_ohe = pd.get_dummies(df_clean['deposit_type'], prefix='deposit_type')
df_clean = df_clean.join(deposit_ohe)

# Hapus kolom tidak terpakai
df_clean.drop('deposit_type', axis=1)

df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
0                     0     PRT         Direct               Direct   

   is_repeated_guest  previous_cancellations  ... is_have_kids meal_FB  \
0                  0                       0  ...            0       0   

   meal_HB meal_SC    negara  is_lokal  room_as_ordered  \
0        0       0  Portugal         1                1   

  deposit_type_No Deposit  deposit_type_Non Refund  deposit_type_Refundable  
0                       1                        0                        0  

[1 rows x 43 columns]

# Agent dan Company

Berisi id dari agent travel dan company

In [40]:
df_clean[['agent', 'company']]

agent  company
0         NaN      NaN
1         NaN      NaN
2         NaN      NaN
3       304.0      NaN
4       240.0      NaN
...       ...      ...
119385  394.0      NaN
119386    9.0      NaN
119387    9.0      NaN
119388   89.0      NaN
119389    9.0      NaN

[118987 rows x 2 columns]

In [41]:
df_clean[['agent', 'company']].isnull().sum()

agent       16263
company    112221
dtype: int64

Banyak nilai kosong pada kedua kolom

Karena berisi id, nilai kosong akan diubah menjadi 0

In [42]:
df_clean[['agent', 'company']] = df_clean[['agent', 'company']].fillna(0)

In [43]:
df_clean[['agent', 'company']].isnull().sum()

agent      0
company    0
dtype: int64

In [44]:
df_clean['agent'].value_counts()
df_clean['company'].value_counts()

0.0      112221
40.0        924
223.0       784
67.0        267
45.0        247
          ...  
142.0         1
32.0          1
11.0          1
487.0         1
376.0         1
Name: company, Length: 349, dtype: int64

Karena nilainya sangat beragam, kedua kolom akan dihapus

In [45]:
df_clean = df_clean.drop(['agent', 'company'], axis=1)

# Customer Type

Perlu dilakukan OHE pada kolom customer_type dengan merubah setiap nilai kolom menjadi kolom baru dengan nilai biner yaitu 0 atau 1. Penggunaan OHE ini karena customer_type tidak memiliki urutan/tingkatan.

In [46]:
customer_ohe = pd.get_dummies(df_clean['customer_type'], prefix='customer_type')
df_clean = df_clean.join(customer_ohe)

# Hapus kolom tidak terpakai
df_clean.drop('customer_type', axis=1)

df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
0                     0     PRT         Direct               Direct   

   is_repeated_guest  previous_cancellations  ...    negara is_lokal  \
0                  0                       0  ...  Portugal        1   

   room_as_ordered deposit_type_No Deposit  deposit_type_Non Refund  \
0                1                       1                        0   

  deposit_type_Refundable  customer_type_Contract  customer_type_Group  \
0                       0                       0                    0   

   customer_type_Transient customer_type_Transient-Party  
0                        1                             0  

[1 rows x 45 columns]

# ADR

Menormalisasi ADR dengan mengeluarkan Outlier

In [47]:
# Batas atas dan bawah ADR
adr_Q1 = df['adr'].quantile(0.25)
adr_Q3 = df['adr'].quantile(0.75)
adr_IQR = adr_Q3 - adr_Q1
adr_low_limit = adr_Q1 - (adr_IQR * 1.5)
adr_high_limit = adr_Q3 + (adr_IQR * 1.5)

print(adr_IQR)
print(adr_low_limit)
print(adr_high_limit)

56.709999999999994
-15.774999999999991
211.065


In [62]:
# # Mengeluarkan adr outlier dan data 0 dan minus
# dfx=df.query("`adr` > 0")
# dfz=dfx.query("`adr` < 211")
df[ (df['adr']<0) & (df['adr']>211)].value_counts()

Series([], dtype: int64)

In [49]:
df_clean = df_clean[ (df_clean['adr']>0) & (df_clean['adr']<211)]

In [50]:
df_clean.head(1)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
2            0          7                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
2                     1     GBR         Direct               Direct   

   is_repeated_guest  previous_cancellations  ...  negara is_lokal  \
2                  0                       0  ...  Others        0   

   room_as_ordered deposit_type_No Deposit  deposit_type_Non Refund  \
2                0                       1                        0   

  deposit_type_Refundable  customer_type_Contract  customer_type_Group  \
2                       0                       0                    0   

   customer_type_Transient customer_type_Transient-Party  
2                        1                             0  

[1 rows x 45 columns]

# Total_of_special_requests, required_car_parking_spaces, dan days_in_waiting_list

- Kolom total_of_special_requests, required_car_parking_spaces, dan days_in_waiting_list sudah memiliki value numerical, akan tetapi untuk memudahkan maka diubah medengan menggunakan conditions dengan asumsi:
- Jika customer memiliki permintaan khusus total_of_special_requests = 0 maka dikategorikan sebagai 0, sedangkan jika customer memiliki permintaan khusus >= 1 maka dikategorikan sebagai 1. Hal ini berlaku juga untuk kolom required_car_parking_spaces dan days_in_waiting_list.

In [51]:
df_clean['new_total_special_request'] = [0 if x == 0 else 1 for x in df_clean['total_of_special_requests']]

df_clean['new_required_car_parking_spaces'] = [0 if x == 0 else 1 for x in df_clean['required_car_parking_spaces']]

df_clean['new_days_in_waiting_list'] = [0 if x == 0 else 1 for x in df_clean['days_in_waiting_list']]



In [52]:
# Hapus kolom tidak terpakai
df_clean.drop(['total_of_special_requests', 'required_car_parking_spaces', 'days_in_waiting_list'], axis=1)

df_clean.head(3)

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
2            0          7                        27                        0   
3            0         13                        27                        0   
4            0         14                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
2                     1     GBR         Direct               Direct   
3                     1     GBR      Corporate            Corporate   
4                     2     GBR      Online TA                TA/TO   

   is_repeated_guest  previous_cancellations  ... deposit_type_No Deposit  \
2                  0                       0  ...                       1   
3                  0                       0  ...                       1   
4                  0                       0  ...                       1   

  deposit_type_Non Refund  deposit_type_Refundable customer_type_Contract  \
2                       0                        0                      0   
3                       0                        0                      0   
4                       0                        0                      0   

   customer_type_Group customer_type_Transient  customer_type_Transient-Party  \
2                    0                       1                              0   
3                    0                       1                              0   
4                    0                       1                              0   

   new_total_special_request  new_required_car_parking_spaces  \
2                          0                                0   
3                          0                                0   
4                          1                                0   

  new_days_in_waiting_list  
2                        0  
3                        0  
4                        0  

[3 rows x 48 columns]

# Reservation_status

Redundant dengan kolom `is_canceled`, akan dihapus

In [53]:
df_clean = df_clean.drop('reservation_status', axis=1)

# Reservation_status_date

Kolom akan di hapus

In [54]:
df_clean = df_clean.drop('reservation_status_date', axis=1)

# Name, email, phone-number, credit_card

Kolom akan dihapus karena nilai masing-masing kolom sangat unik

In [55]:
df_clean = df_clean.drop(['name', 'email', 'phone-number', 'credit_card'], axis=1)

# Dataset akhir setelah Preprocessing

In [56]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 113323 entries, 2 to 119389
Data columns (total 42 columns):
 #   Column                           Non-Null Count   Dtype         
---  ------                           --------------   -----         
 0   is_canceled                      113323 non-null  int64         
 1   lead_time                        113323 non-null  int64         
 2   arrival_date_week_number         113323 non-null  int64         
 3   stays_in_weekend_nights          113323 non-null  int64         
 4   stays_in_week_nights             113323 non-null  int64         
 5   country                          113323 non-null  object        
 6   market_segment                   113323 non-null  object        
 7   distribution_channel             113323 non-null  object        
 8   is_repeated_guest                113323 non-null  int64         
 9   previous_cancellations           113323 non-null  int64         
 10  reserved_room_type               113323 non-

In [57]:
df_clean.head()

is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
2            0          7                        27                        0   
3            0         13                        27                        0   
4            0         14                        27                        0   
5            0         14                        27                        0   
6            0          0                        27                        0   

   stays_in_week_nights country market_segment distribution_channel  \
2                     1     GBR         Direct               Direct   
3                     1     GBR      Corporate            Corporate   
4                     2     GBR      Online TA                TA/TO   
5                     2     GBR      Online TA                TA/TO   
6                     2     PRT         Direct               Direct   

   is_repeated_guest  previous_cancellations  ... deposit_type_No Deposit  \
2                  0                       0  ...                       1   
3                  0                       0  ...                       1   
4                  0                       0  ...                       1   
5                  0                       0  ...                       1   
6                  0                       0  ...                       1   

  deposit_type_Non Refund  deposit_type_Refundable customer_type_Contract  \
2                       0                        0                      0   
3                       0                        0                      0   
4                       0                        0                      0   
5                       0                        0                      0   
6                       0                        0                      0   

   customer_type_Group customer_type_Transient  customer_type_Transient-Party  \
2                    0                       1                              0   
3                    0                       1                              0   
4                    0                       1                              0   
5                    0                       1                              0   
6                    0                       1                              0   

   new_total_special_request  new_required_car_parking_spaces  \
2                          0                                0   
3                          0                                0   
4                          1                                0   
5                          1                                0   
6                          0                                0   

   new_days_in_waiting_list  
2                         0  
3                         0  
4                         0  
5                         0  
6                         0  

[5 rows x 42 columns]